add Arguments Optional

rosannaschaefer opened the pull request:
Hi Ben, this is the Arguments Optional challenge, here are the instructions:

> Create a function that sums two arguments together. If only one argument is provided, then return a function that expects one argument and returns the sum.
> For example, addTogether(2, 3) should return 5, and addTogether(2) should return a function.
> Calling this returned function with a single argument will then return the sum:
> var sumTwoAnd = addTogether(2);
> sumTwoAnd(3) returns 5.
> If either argument isn't a valid number, return undefined.

This is my fourth try and I keep hitting the same roadblock. The first three tries are commented out below the code in case you want to see. Basically I've gotten this far: if there are more than one argument, and both are numbers: add them. When there's a test such as `addTogether(2)(3)` I can't figure out how to access the `(3)`. So when there is only one argument (`else if (arguments.length === 1)`) I can't figure out where to go from there.

The instructions say "If only one argument is provided, then return a function that expects one argument and returns the **sum**." but the sum of what?

I really tried to challenge myself and not look at the solutions, because I feel very close to the answer.

rosannaschaefer committed:
add Arguments Optional

benschac commented:
Can you try again? Looking at your comment but there is no code in the PR.  Happy to walk through the branching process.
<img width="1033" alt="screen shot 2018-07-13 at 9 10 20 am" src="https://user-images.githubusercontent.com/2502947/42693288-ab7c6de6-867c-11e8-92de-8da9b813c38a.png">


DO NOT DELETE

rosannaschaefer committed:
saved this time

rosannaschaefer commented:
I think I just forgot to save it in my text editor first. My bad.

benschac commented on arguments-optional.js:
```diff
@@ -0,0 +1,79 @@
+// If either argument isn't a valid number, return undefined.
+
+function addTogether() {
+  
```
Kill unneed space

benschac commented on arguments-optional.js:
```diff
@@ -0,0 +1,79 @@
+
+function addTogether() {
+  
+  if (arguments.length > 1 && typeof(arguments[0], arguments[1]) === "number") {
```
Why did you not add a number in the parameter brackets?

```js
function addTogether(number1) {
...
}
```

You then don't need the arguments object.  Just test if the `number1` parameter is undefined

Better yet, use a default parameterhttps://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Default_parameters

```js
if (number1 === null) // do your logic
```

benschac commented on arguments-optional.js:
```diff
@@ -0,0 +1,79 @@
+// sumTwoAnd(3) returns 5.
+
+// If either argument isn't a valid number, return undefined.
+
```
use JS DOC: http://usejsdoc.org/about-getting-started.html
like this:
```js
/**
 * sums two arguments together. If only one argument is provided, then return a function that expects one argument and returns the sum.
 * 
 * @param {number} str to translate from binary
 * 
 * @return {string} translated string to enlish 
 */
```

benschac commented:
Please comment the URL of the problem you're working on so I can reference it easily.

- Additionally, create a `.gitignore` file and add `.DS_STORE` https://stackoverflow.com/questions/107701/how-can-i-remove-ds-store-files-from-a-git-repository

benschac reviewed (commented):
good start

DO NOT DELETE

benschac commented:
More on closures.  THEY ARE VERY IMPORTANT. TAKE YOU TIME WITH THEM:
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Closures
http://javascriptissexy.com/understand-javascript-closures-with-ease/

benschac commented:
DO NOT MERGE UNTIL I LGTM

rosannaschaefer committed:
solved algorithm, added JSDoc, added .gitignore

rosannaschaefer committed:
solved algorithm, added JSDoc, added .gitignore

rosannaschaefer commented:
Hi Ben, I was able to get to a solution. A couple of things:

- I accidentally added two commits because I was playing around with using the embedded Git control in VS Code and trying to do it just in the terminal. Best way to learn is to try right?
- I feel comfortable with closures and was able to use it to get to the solution.
- I learned what JSDoc is and added it in. Let me know if it looks correct to you since it's my first time using it.
- You told me to create a .gitignore file (I added this) and _add_ .DS_STORE but sent a link that shows how to remove .DS_STORE. I googled it and it looks as though everyone is trying to remove it so I'm assuming you want me to remove it but I figured I'd check first.

benschac commented on arguments-optional.js:
```diff
@@ -0,0 +1,22 @@
+/**
+ *sums two arguments together. If only one argument is provided, returns a function that expects one argument and returns the sum. If either argument isn't a valid number, returns undefined.
+ *
+ * @param {*} a
```
```js
@param {number} a number to add
@param {number} b number to add
```

rosannaschaefer replied:
the parameter can be any type, not just a number.

benschac replied:
than `@param{any}`

rosannaschaefer replied:
<img width="449" alt="screen shot 2018-07-15 at 6 37 57 pm" src="https://user-images.githubusercontent.com/26444164/42743304-4b6cf7f2-885e-11e8-8ff7-21e81798bb1b.png">

going by the documentation..

benschac replied:
http://usejsdoc.org/tags-param.html

Look at you digging into the documentation calling my bullshit! Great job! +100 points

benschac commented on arguments-optional.js:
```diff
@@ -0,0 +1,22 @@
+ *
+ * @param {*} a
+ * @param {*} b
+ * @returns {number|undefined} sum of a and b or undefined
```
```js
@returns {number|function} sum of a and b or function
```

benschac commented on arguments-optional.js:
```diff
@@ -0,0 +1,22 @@
+/**
```
I need the URL of the code challenge so I don't have to go searching for it.

benschac replied:
always think about how to make it easier for your reviewer to look at your code.  More context is always better.

rosannaschaefer replied:
https://learn.freecodecamp.org/javascript-algorithms-and-data-structures/intermediate-algorithm-scripting/arguments-optional 

Will include that from the get-go from now on.

benschac replied:
Can you update the code here too?

benschac commented on arguments-optional.js:
```diff
@@ -0,0 +1,22 @@
+ * @param {*} b
+ * @returns {number|undefined} sum of a and b or undefined
+ */
+function addTogether(a, b) {
```
use default parameters: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Default_parameters
```js
function addTogether(a=null, b=null) {
...
}
if (a !== null && b !== null) {
   
}

benschac commented on arguments-optional.js:
```diff
@@ -0,0 +1,22 @@
+    return a + b
+  } 
+  else if (arguments.length === 1 && typeof(a) === "number") {
+    return function addOther(c) {
```
not sure why you need a `c` variable here.

rosannaschaefer replied:
"If only one argument is provided, then return a function that expects one argument and returns the sum...Calling this returned function with a single argument will then return the sum" 

I did include in the description in the JSDoc part. Is that the right place for it?

benschac replied:
You don't need `c` variable. Look at code posted below.

benschac commented on arguments-optional.js:
```diff
@@ -0,0 +1,22 @@
+ * @returns {number|undefined} sum of a and b or undefined
+ */
+function addTogether(a, b) {
+  if (arguments.length === 2 && typeof(a, b) === "number") {
```
I still don't not why you need the `arguments` object in this function.

rosannaschaefer replied:
Whether there are one or two arguments is a condition of the code challenge. I used `arguments.length` to check whether there is one argument or two.

benschac replied:
wrote a comment below.  You don't need arguments object.

benschac commented on arguments-optional.js:
```diff
@@ -0,0 +1,22 @@
+    }
+  } 
+  else {
+    return undefined
```
is returning `undefined` a requirement of the solution?

rosannaschaefer replied:
https://learn.freecodecamp.org/javascript-algorithms-and-data-structures/intermediate-algorithm-scripting/arguments-optional

yes.

rosannaschaefer replied:
Also "If either argument isn't a valid number, returns undefined." is in the description in the JSDoc part. Am I putting that in the right place?

benschac replied:
could also do `@return void`

benschac reviewed (changes requested):
good progress, keep it up.

benschac commented:
I need to look at the problem spec and make sure I'm following the requirements.

```js
function addTogether(a=0) {
  return function (b=0) {
    return a + b;
  }
}

// or the es6 way
const addTogether = (a=0) => (b=0) => a + b;
```

benschac commented:
`git rm -rf .DS_STORE` should work. then add and commit the change.

benschac commented:
VSCode git is v. good. Sometimes it doesn't work. just make sure you can commit code and you're good.

benschac commented:
```js
function addTogether(a, b) {
  if (typeof a === 'number' && typeof b === 'number') {
    return a + b;
  }

  if (typeof a !== 'number') {
    return undefined;
  }

  if (typeof b === 'string') {
    return undefined;
  }


  return function (b) {
    return Array.isArray(b) ? undefined : a + b;
  }
}
// es6
function addTogether(a, b) {
  if (typeof a === 'number' && typeof b === 'number') {
    return a + b;
  }

  if (typeof a !== 'number') return undefined;

  if (typeof b === 'string') return undefined;


  return b =>  Array.isArray(b) ? undefined : a + b;
}

addTogether(2,3);
```

rosannaschaefer commented:
One thing I can't figure out about your solution is how it work on `addTogether(2)(3)`. Can you explain that?

EDIT: never mind, calling the anonymous function(b) brings the second argument in `addTogether(2)(3)` into existence.

benschac commented:
Sure, that's a closure. A function returning a function. 
```js
return b =>  Array.isArray(b) ? undefined : a + b;
```
it's an arrow function. same as `function(){...}`.  Here I'm testing if `b` is an array.  If not than we return the second function that returns `a+b`. If you're still confused throw in some console.logs and try to get some insight. 

Happy to chat if that doesn't work.

rosannaschaefer committed:
delete .DS_STORE,  solution improvements

rosannaschaefer committed:
delete .DS_STORE, solution improvements

benschac commented on arguments-optional.js:
```diff
@@ -0,0 +1,16 @@
+ * @returns {number|undefined} sum of a and b or undefined
+ */
+function addTogether(a, b) {
+  if (typeof(a, b) === "number") return a + b;
```
@rosannaschaefer I didn't know `typeof` could take two params! learned something new!

benschac commented on arguments-optional.js:
```diff
@@ -0,0 +1,16 @@
+function addTogether(a, b) {
+  if (typeof(a, b) === "number") return a + b;
+
+  if (arguments.length === 1 && typeof(a) === "number") 
```
I still don't think you need the `arguments` object, but that's okay.

rosannaschaefer replied:
I did try it without it and it didn't work

benschac reviewed (approved):
LGTM  🎉 Trusting you checked this in the console on line and tests are passing!

rosannaschaefer commented:
thanks @benschac 🎉
